custom/yeksi_naa_reports/sqldata.py

```diff
@@ -809,6 +809,87 @@ def headers(self):
         return headers
 
 
+class SatisfactionRateAfterDeliveryData(VisiteDeLOperateurPerProductDataSource):
+    slug = 'satisfaction_rate_after_delivery'
+    comment = '% products ordered vs. delivered'
+    title = 'Satisfaction Rate after delivery'
+    show_total = True
+    custom_total_calculate = True
+
+    def calculate_total_row(self, products):
+        total_row = ['Total (CFA)']
+        for i in range(len(self.months)):
+            total_row.append(self.percent_fn(
+                sum(
+                    products[product_id][i]['numerator'] for product_id in products if
+                    products[product_id][i]['numerator'] and products[product_id][i]['denominator']
+                ),
+                sum(
+                    products[product_id][i]['denominator'] for product_id in products if
+                    products[product_id][i]['numerator'] and products[product_id][i]['denominator']
+                )
+            ))
+        return total_row
+
+    @property
+    def group_by(self):
+        return ['real_date_repeat', 'product_id', 'product_name']
+
+    @property
+    def columns(self):
+        columns = [
+            DatabaseColumn("Date", SimpleColumn('real_date_repeat')),
+            DatabaseColumn("Product ID", SimpleColumn('product_id')),
+            DatabaseColumn("Product Name", SimpleColumn('product_name')),
+            DatabaseColumn("Quantity of the product delivered", SumColumn('amt_delivered_convenience')),
+            DatabaseColumn("Quantity of the product  suggested", SumColumn('ideal_topup')),
+        ]
+        return columns
+
+    @property
+    def rows(self):
+        rows = self.get_data()
+        data = {}
+        product_names = {}
+        for row in rows:
+            if self.months[0] <= row['real_date_repeat'] < self.months[-1] + relativedelta(months=1):
```

Comment:
nit: it would be much easier to parse this if it was extracted into a function. e.g. `if self._date_is_in_range(row['real_date_repeat'])` (or whatever an appropriate name would be)